google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/UnitOfWork.java

```diff
@@ -142,6 +143,18 @@ public boolean isActive() {
   /** @return the commit timestamp of this transaction or null if there is no commit timestamp. */
   Timestamp getCommitTimestampOrNull();
 
+  /**
+   * @return the {@link CommitResponse} of this transaction. Will throw a {@link SpannerException}
+   *     if there is no {@link CommitResponse}.
+   */
+  CommitResponse getCommitResponse();
+
+  /**
+   * @return the {@link CommitResponse} of this transaction or null if there is no {@link
+   *     CommitResponse}.
+   */
+  CommitResponse getCommitResponseOrNull();
```

Comment:
You don't need both these methods. Even if these methods are left here, all we need is a single getCommitResponse method that returns a commit response or null.

Reply:
This is not intended for public use (it's a package private interface). The reason that there are two different versions is that some APIs require a method to throw an error if the connection / transaction is not in a valid state for a specific method call, and in other cases it expects it to return `null`.

Reply:
Which APIs? It sounds like the responsibility for interpreting and handling a null return or an exception belongs in those methods.

Reply:
The two APIs being served are:
1. The `Connection` interface with its `getCommitResponse()` and other similar methods. These all throw an exception if they are called while the `Connection` is in an invalid state for the method.
2. The interpreted SQL interface which will return a `ResultSet` with a `null` value if there is no commit response available at that moment.

The `Connection` interface could use the `getCommitResponseOrNull()` method, check for a null value and then throw an exception. The disadvantage of that would be that the `Connection` does not know why the value is null, and can therefore only throw an exception with a generic message. By delegating the exception that should be thrown to the `UnitOfWork`, the `UnitOfWork` can throw an exception with a more specific message indicating why there is no commit response.